### PR TITLE
Finally added support for tate configs.

### PR DESCRIPTION
**Pull request by D-Hiatt**
Everything should still work the same way.  Load a `spec_config` object by using `with open(path, 'r') as f:
            return json.load(f, cls=spec_decoder)`

It should be able to decode both the old `spec_config` json format as well as the new `tate_config` format.  Additionally, there is now a `spec_config.type` member that controls which type of json to save it as.  Use `spec_config.switch_type()` to switch between 'spec_config' and 'tate_config'.

The objects in `object.py` do not keep track of translations or annotations.  Instead, they are converted to their relevant `prop_item` counterpart.  

Here's a quick breakdown on each object, their members, and helpful functions:

- spec_config
    - self.switch_type() : Toggles between saving as tate_config and spec_config
    - self.save(path) : Saves the json file to the path
    - self.type : Will either be 'spec_config' or 'tate_config'.  Use self.switch_type() to toggle
    - self.runs : A list containing `spec_run` objects.
-  spec_run
    - Contains a whole bunch of run related variables.
    - self.properties : A `props` object.  This should never be overwritten.
- props
   - Always contain all possible SPECjbb properties.
   - self.writeconfig(path) : Writes the specjbb.props file to the path
   - self.get_modified() : Returns all modified specjbb properties in this run
   - self.get_all() : Returns all possible specjbb properties
   - self.set(key, value) : Can be used to set a `prop_item` with the specified key.  The key should match the actual property name, ie 'specjbb.controller.type'
- prop_item
    - self.prop : The full property name, ie 'specjbb.controller.type'
    - self.desc : The annotation corresponding to this property
    - self.def_value : The default value
    - self.value : The current value
    - self.value_validator : A function that returns true if the argument provided is valid for this property
    - self.input_validator : Can be used to clean up user input.  Its a function that takes a single char as an argument, and returns true if it is valid.  Prevents users from entering letters when a number is expected, etc.
    - self.valid_opts : A list containing valid property values, or `None` if value range is continuous
    - self.help_text : A message that informs the user why their selection was rejected by `self.value_validator`
    - self.set(value) : Checks if the value is valid, then sets `self.value`
   - self.reset() : Resets `self.value` to `self.def_value`

### CODE REVIEW

File: objects.py [hunk 1]
```diff
@@ -43,13 +97,46 @@ def __init__(self, *args, **kwargs):
...
+    def decode_tate(self, obj, templates):
+        runlist = obj.get('RunList', '')
+        if (runlist == ''):
+            raise AssertionError("Json file has no run lists")
+        #Run lists is a list, templates is a dict.
+        ret = spec_config()
+        ret.type = "tate_config"
+        for run in runlist:
+            r = spec_run()
+            template_type = run.get('template_type', '')
```
Comment (benjspriggs) [inline objects.py]: Template type is required, so if this isn't provided you could also just throw an exception:
```python
try:
    run.get("template_type")
except IndexError:
    raise AssertionError("it's gone")
```

File: objects.py [hunk 2]
```diff
@@ -99,7 +210,24 @@ def tojson(self):
             "report_level": self.report_level,
             "skip_report": self.skip_report,
             "ignore_kit_validation": self.ignore_kit_validation,
-            "props": self.properties.tojson()
+            "props": self.properties._tojson()
+        }
+
+    def _totateconfig(self):
```
Comment (benjspriggs) [inline objects.py]: We should test this using [`unittest`](https://docs.python.org/3/library/unittest.html#basic-example).

File: speccurses.py [hunk 3]
```diff
@@ -9,19 +9,8 @@
...
-LEFT = 260
-RIGHT = 261
-HOME = 262
-BACK = 263
-DEL = 330
-PGDN = 338
-PGUP = 339
-END = 360
+#KEY_ENTER = 343, but my ENTER key registers as 10 =/
+KEY_ENTER = 10
```
Comment (benjspriggs) [inline speccurses.py]: Are we looking at enter-key issues with different keyboards? Is there a different code entered for different systems?

File: speccurses.py [hunk 4]
```diff
@@ -448,17 +457,22 @@ def _handle(msg):
...
+           #     pad.refresh(c + 1, 0, cury, xoffset, cury + c, width - xoffset - 1)
+       # else:
           #  if c + cury < maxy:
           #      pad.refresh(0, 0, cury, xoffset, cury + c, width - xoffset - 1)
            # else:
                 #print 0 -> c
                 #print logmax - (c
-                pad.refresh(0,0, maxy - c, xoffset, maxy, width - xoffset - 1)
-                pad.refresh(logmax - c, 0, cury, xoffset, maxy - c - 1, width - xoffset - 1)
+        #        pad.refresh(0,0, maxy - c, xoffset, maxy, width - xoffset - 1)
```
Comment (benjspriggs) [inline speccurses.py]: We should remove unused code.

File: speccurses.py [hunk 5]
```diff
@@ -492,7 +506,7 @@ def _handle(msg):
     for r in config.runs:
         # draw_status_bar(stdscr, "Starting run '{}'".format(r.tag))
 
-        result = r.run(path, _handle)
+        result = r.run(path, _handle, lambda x:x)
```
Comment (benjspriggs) [inline speccurses.py]: If we're passing an additional argument to handle errors, shouldn't we pass something other than this? Should this be the default argument for `run()`?

Comment (D-Hiatt) [reply speccurses.py]: Both stdout and stderr are being written to log files already, so it’s not crucial the text be shown in console as well.  Specjbb uses stderr for a bunch of status messages and clutters things up I think.

File: speccurses.py (hunk 3 above)
Comment (D-Hiatt) [reply speccurses.py]: 10 is ascii for \n. Curses doesn’t seem to pick up the actual enter key

**Pull request merged**